WIP: [GSoC] GPU-enabled OpenCV.js, add webgpu-dawn support for dnn module.

NALLEIN opened the pull request:
### Overview

- Proposal : [GPU-enabled OpenCV.js](https://summerofcode.withgoogle.com/dashboard/project/5352882720735232/details/)

- Mentor : Zhiwen Wu @wzw-intel 

- Student : Hao Jiang [@NALLEIN](https://github.com/NALLEIN)
### This pull request changes

#### **Done**

- Use [Dawn , a  webgpu implemention ](https://dawn.googlesource.com/dawn)to implement some operators under the opencv/modiles/dnn module, refer to [the implementation of Vulkan](https://github.com/opencv/opencv/pull/12703) and reuse the shader. Passed the corresponding test cases on ubuntu.

```
git clone https://github.com/opencv/opencv_extra.git
export OPENCV_TEST_DATA_PATH=${PATH_TO}/opencv_extra/testdata

sudo apt-get install libvulkan1 mesa-vulkan-drivers vulkan-utils
git clone https://github.com/NALLEIN/opencv.git
cd opencv/
git checkout -b layerTest origin/layerTest
mkdir build
cd build
cmake -D CMAKE_BUILD_TYPE=Release -DWITH_WEBGPU=ON -D CMAKE_INSTALL_PREFIX=/usr/local ..
make -j8
$(PATH_TO_OPENCV)/build/bin/opencv_test_dnn --gtest_filter=*layers*
```

- Use [emscripten](https://emscripten.org/) to compile Dawn's API into a wasm format that can be executed in the chrome browser. This module can be compiled to JavaScript using `--build_webgpu` flag, E.g ``` python ./platforms/js/build_js.py build_js --build_webgpu```.
Use the latest [emsdk](https://emscripten.org/docs/getting_started/downloads.html)  version 1.40.1, you can refer to [opencv.js tutorials ](https://docs.opencv.org/3.4/d4/da1/tutorial_js_setup.html)to set up the environment.

####**To do** 

- Verify wheher the DNN_BACKEND_WGPU compiled by emscripten can run correctly in chrome.

- Write the unit test and perf test for DNN_BACKEND_WGPU.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under OpenCV (BSD) License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

alalek commented:
> .so

Pre-compiled binary blobs are not allowed in repository with source code.

Provide script of their compilation instead (for example, as part of build_js.py).

huningxin commented:
For OpenCV.js WebGPU implementation, it doesn't require the native .so files. If they are only used for testing purpose, please exclude them in this PR.

wzw-intel commented:
@NALLEIN , please fix the merge conflicts

huningxin commented:
@NALLEIN , I tested the dawn path. There are two typos in your [comment](https://github.com/opencv/opencv/pull/18066#issue-465496038)

> opencv/modiles/dnn

It should be `opencv/modules/dnn`.

> git checkout -b layerTest origin/layerTest

It should be `gsoc_2020_webgpu` branch.

The test results of `opencv_test_dnn --gtest_filter=*layers*` are
```
[==========] 388 tests from 7 test cases ran. (2554 ms total)
[  PASSED  ] 382 tests.
[  FAILED  ] 6 tests, listed below:
[  FAILED  ] Test_Darknet_layers.convolutional/0, where GetParam() = WEBGPU/DAWN
[  FAILED  ] Test_ONNX_layers.Constant/0, where GetParam() = WEBGPU/DAWN
[  FAILED  ] Test_TensorFlow_layers.concat_3d/0, where GetParam() = WEBGPU/DAWN
[  FAILED  ] Test_TensorFlow_layers.slim_batch_norm/0, where GetParam() = WEBGPU/DAWN
[  FAILED  ] Test_TensorFlow_layers.ave_pool_same/0, where GetParam() = WEBGPU/DAWN
[  FAILED  ] Test_Torch_layers.net_inception_block/0, where GetParam() = WEBGPU/DAWN
```

Did you happen to investigate the failures?

huningxin commented:
I also tried the emscripten path, when run perf_conv2d.html, I got error as
```
opencv.js loaded
DevTools failed to load SourceMap: Could not load content for wasm://wasm/{{{ FILENAME_REPLACEMENT_STRINGS_WASM_BINARY_FILE }}}.map: HTTP error: status code 404, net::ERR_UNKNOWN_URL_SCHEME
perf_conv2d.js:85 Time cost(ms) : 20.024999997986015
perf_conv2d.js:86 null
perf_conv2d.js:98 Time1 cost(ms) : 6.267475247519128
perf_conv2d.js:99 Uint8Array(406800) [171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, …]
perf_conv2d.js:100 Net compute succeed
opencv.js:1808 undefined
03277a66:0x980883 Uncaught (in promise) RuntimeError: memory access out of bounds
    at dlmalloc (<anonymous>:wasm-function[14841]:0x980883)
    at operator new(unsigned long) (<anonymous>:wasm-function[14685]:0x979d69)
    at std::__2::__libcpp_allocate(unsigned long, unsigned long) (<anonymous>:wasm-function[14493]:0x977d49)
    at std::__2::allocator<char>::allocate(unsigned long, void const*) (<anonymous>:wasm-function[14508]:0x977f92)
    at std::__2::allocator_traits<std::__2::allocator<char> >::allocate(std::__2::allocator<char>&, unsigned long) (<anonymous>:wasm-function[14507]:0x977e9c)
    at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::__init(char const*, unsigned long) (<anonymous>:wasm-function[14702]:0x97a4e8)
    at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::basic_string(std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&) (<anonymous>:wasm-function[14699]:0x97a37b)
    at cv::Exception::Exception(cv::Exception const&) (<anonymous>:wasm-function[3405]:0x1b152c)
    at cv::error(cv::Exception const&) (<anonymous>:wasm-function[3403]:0x1b12da)
    at cv::error(int, std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&, char const*, char const*, int) (<anonymous>:wasm-function[3396]:0x1b0c0a)
dlmalloc @ 03277a66:0x980883
operator new(unsigned long) @ 03277a66:0x979d69
std::__2::__libcpp_allocate(unsigned long, unsigned long) @ 03277a66:0x977d49
std::__2::allocator<char>::allocate(unsigned long, void const*) @ 03277a66:0x977f92
std::__2::allocator_traits<std::__2::allocator<char> >::allocate(std::__2::allocator<char>&, unsigned long) @ 03277a66:0x977e9c
std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::__init(char const*, unsigned long) @ 03277a66:0x97a4e8
std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::basic_string(std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&) @ 03277a66:0x97a37b
cv::Exception::Exception(cv::Exception const&) @ 03277a66:0x1b152c
cv::error(cv::Exception const&) @ 03277a66:0x1b12da
cv::error(int, std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&, char const*, char const*, int) @ 03277a66:0x1b0c0a
cv::setSize(cv::Mat&, int, int const*, unsigned long const*, bool) @ 03277a66:0x110f64
cv::Mat::copySize(cv::Mat const&) @ 03277a66:0x112b4f
cv::dnn::dnn4_v20200310::Net::Impl::getBlob(cv::dnn::dnn4_v20200310::LayerPin const&) @ 03277a66:0x4a6a0c
cv::dnn::dnn4_v20200310::Net::forward(std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&) @ 03277a66:0x4a5134
Wrappers::dnn_Net_forward_wrapper_1(cv::dnn::dnn4_v20200310::Net&) @ 03277a66:0x471ad
emscripten::internal::FunctionInvoker<cv::Mat (*)(cv::dnn::dnn4_v20200310::Net&), cv::Mat, cv::dnn::dnn4_v20200310::Net&>::invoke(cv::Mat (**)(cv::dnn::dnn4_v20200310::Net&), cv::dnn::dnn4_v20200310::Net*) @ 03277a66:0x7702b
dynCall_iii @ 03277a66:0x9ae3b7
ret.<computed> @ opencv.js:8649
(anonymous) @ opencv.js:8726
Promise.then (async)
(anonymous) @ opencv.js:8775
handleSleep @ opencv.js:8712
handleAsync @ opencv.js:8773
readBufferAsync @ opencv.js:1808
cv::dnn::webgpu::Buffer::MapReadAsyncAndWait() @ 03277a66:0x6982ec
cv::dnn::webgpu::Tensor::mapRead() @ 03277a66:0x6a1487
cv::dnn::copyToMat(cv::Mat&, cv::dnn::webgpu::Tensor&) @ 03277a66:0x5d4e8d
cv::dnn::WGPUBackendWrapper::copyToHost() @ 03277a66:0x5d68fd
cv::dnn::dnn4_v20200310::Net::Impl::getBlob(cv::dnn::dnn4_v20200310::LayerPin const&) @ 03277a66:0x4a6759
cv::dnn::dnn4_v20200310::Net::forward(std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&) @ 03277a66:0x4a5134
Wrappers::dnn_Net_forward_wrapper_1(cv::dnn::dnn4_v20200310::Net&) @ 03277a66:0x471ad
emscripten::internal::FunctionInvoker<cv::Mat (*)(cv::dnn::dnn4_v20200310::Net&), cv::Mat, cv::dnn::dnn4_v20200310::Net&>::invoke(cv::Mat (**)(cv::dnn::dnn4_v20200310::Net&), cv::dnn::dnn4_v20200310::Net*) @ 03277a66:0x7702b
dynCall_iii @ 03277a66:0x9ae3b7
ret.<computed> @ opencv.js:8649
Module.dynCall_iii @ opencv.js:9033
dynCall_iii_1083 @ VM9080:4
dnn_Net$forward @ VM9082:8
proto.<computed> @ opencv.js:5754
testCaffeLayer @ perf_conv2d.js:83
async function (async)
testCaffeLayer @ perf_conv2d.js:75
runButton.onclick @ perf_conv2d.js:122
```

huningxin commented on modules/dnn/src/webgpu/src/buffer.cpp:
```diff
@@ -0,0 +1,122 @@
+            await gpuReadBuffer.mapAsync(GPUMapMode.READ);
+            const arrayBuffer = gpuReadBuffer.getMappedRange();
+            console.log(arrayBuffer.length);
+            return arrayBuffer;
```
The `ArrayBuffer` doesn't convert to `void*`. You may need to alloc the memory on Wasm heap and copy the data of `arrayBuffer` to that memory. After use, you need to free that memory in your C code.

Here are some references:
https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#access-memory-from-javascript
https://emscripten.org/docs/api_reference/emscripten.h.html#c.EM_JS

NALLEIN commented:
> @NALLEIN , I tested the dawn path. There are two typos in your [comment](https://github.com/opencv/opencv/pull/18066#issue-465496038)
> 
> > opencv/modiles/dnn
> 
> It should be `opencv/modules/dnn`.
> 
> > git checkout -b layerTest origin/layerTest
> 
> It should be `gsoc_2020_webgpu` branch.
> 
> The test results of `opencv_test_dnn --gtest_filter=*layers*` are
> 
> ```
> [==========] 388 tests from 7 test cases ran. (2554 ms total)
> [  PASSED  ] 382 tests.
> [  FAILED  ] 6 tests, listed below:
> [  FAILED  ] Test_Darknet_layers.convolutional/0, where GetParam() = WEBGPU/DAWN
> [  FAILED  ] Test_ONNX_layers.Constant/0, where GetParam() = WEBGPU/DAWN
> [  FAILED  ] Test_TensorFlow_layers.concat_3d/0, where GetParam() = WEBGPU/DAWN
> [  FAILED  ] Test_TensorFlow_layers.slim_batch_norm/0, where GetParam() = WEBGPU/DAWN
> [  FAILED  ] Test_TensorFlow_layers.ave_pool_same/0, where GetParam() = WEBGPU/DAWN
> [  FAILED  ] Test_Torch_layers.net_inception_block/0, where GetParam() = WEBGPU/DAWN
> ```
> 
> Did you happen to investigate the failures?

There are two failed cases with the same results as the Vulkan backend test.
`Test_TensorFlow_layers.concat_3d` failed because 3D conv operation is not currently supported.
`Test_TensorFlow_layers.ave_pool_same` failed because of the accuracy problem, because the previous Vulkan shader had problems. 

Other cases can be tested successfully but the result is unstable. Sometimes it will fail because the map fails during buffer async reading back. You can verify the other cases separately, for example `./bin/opencv_test_dnn --gtest_filter=*Test_Darknet_layers.convolutional*`. You can test several times, these cases cost a long time (sometimes more than 100ms) and may fail sometimes.

NALLEIN commented:
> I also tried the emscripten path, when run perf_conv2d.html, I got error as
> 
> ```
> opencv.js loaded
> DevTools failed to load SourceMap: Could not load content for wasm://wasm/{{{ FILENAME_REPLACEMENT_STRINGS_WASM_BINARY_FILE }}}.map: HTTP error: status code 404, net::ERR_UNKNOWN_URL_SCHEME
> perf_conv2d.js:85 Time cost(ms) : 20.024999997986015
> perf_conv2d.js:86 null
> perf_conv2d.js:98 Time1 cost(ms) : 6.267475247519128
> perf_conv2d.js:99 Uint8Array(406800) [171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, …]
> perf_conv2d.js:100 Net compute succeed
> opencv.js:1808 undefined
> 03277a66:0x980883 Uncaught (in promise) RuntimeError: memory access out of bounds
>     at dlmalloc (<anonymous>:wasm-function[14841]:0x980883)
>     at operator new(unsigned long) (<anonymous>:wasm-function[14685]:0x979d69)
>     at std::__2::__libcpp_allocate(unsigned long, unsigned long) (<anonymous>:wasm-function[14493]:0x977d49)
>     at std::__2::allocator<char>::allocate(unsigned long, void const*) (<anonymous>:wasm-function[14508]:0x977f92)
>     at std::__2::allocator_traits<std::__2::allocator<char> >::allocate(std::__2::allocator<char>&, unsigned long) (<anonymous>:wasm-function[14507]:0x977e9c)
>     at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::__init(char const*, unsigned long) (<anonymous>:wasm-function[14702]:0x97a4e8)
>     at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::basic_string(std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&) (<anonymous>:wasm-function[14699]:0x97a37b)
>     at cv::Exception::Exception(cv::Exception const&) (<anonymous>:wasm-function[3405]:0x1b152c)
>     at cv::error(cv::Exception const&) (<anonymous>:wasm-function[3403]:0x1b12da)
>     at cv::error(int, std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&, char const*, char const*, int) (<anonymous>:wasm-function[3396]:0x1b0c0a)
> dlmalloc @ 03277a66:0x980883
> operator new(unsigned long) @ 03277a66:0x979d69
> std::__2::__libcpp_allocate(unsigned long, unsigned long) @ 03277a66:0x977d49
> std::__2::allocator<char>::allocate(unsigned long, void const*) @ 03277a66:0x977f92
> std::__2::allocator_traits<std::__2::allocator<char> >::allocate(std::__2::allocator<char>&, unsigned long) @ 03277a66:0x977e9c
> std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::__init(char const*, unsigned long) @ 03277a66:0x97a4e8
> std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::basic_string(std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&) @ 03277a66:0x97a37b
> cv::Exception::Exception(cv::Exception const&) @ 03277a66:0x1b152c
> cv::error(cv::Exception const&) @ 03277a66:0x1b12da
> cv::error(int, std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&, char const*, char const*, int) @ 03277a66:0x1b0c0a
> cv::setSize(cv::Mat&, int, int const*, unsigned long const*, bool) @ 03277a66:0x110f64
> cv::Mat::copySize(cv::Mat const&) @ 03277a66:0x112b4f
> cv::dnn::dnn4_v20200310::Net::Impl::getBlob(cv::dnn::dnn4_v20200310::LayerPin const&) @ 03277a66:0x4a6a0c
> cv::dnn::dnn4_v20200310::Net::forward(std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&) @ 03277a66:0x4a5134
> Wrappers::dnn_Net_forward_wrapper_1(cv::dnn::dnn4_v20200310::Net&) @ 03277a66:0x471ad
> emscripten::internal::FunctionInvoker<cv::Mat (*)(cv::dnn::dnn4_v20200310::Net&), cv::Mat, cv::dnn::dnn4_v20200310::Net&>::invoke(cv::Mat (**)(cv::dnn::dnn4_v20200310::Net&), cv::dnn::dnn4_v20200310::Net*) @ 03277a66:0x7702b
> dynCall_iii @ 03277a66:0x9ae3b7
> ret.<computed> @ opencv.js:8649
> (anonymous) @ opencv.js:8726
> Promise.then (async)
> (anonymous) @ opencv.js:8775
> handleSleep @ opencv.js:8712
> handleAsync @ opencv.js:8773
> readBufferAsync @ opencv.js:1808
> cv::dnn::webgpu::Buffer::MapReadAsyncAndWait() @ 03277a66:0x6982ec
> cv::dnn::webgpu::Tensor::mapRead() @ 03277a66:0x6a1487
> cv::dnn::copyToMat(cv::Mat&, cv::dnn::webgpu::Tensor&) @ 03277a66:0x5d4e8d
> cv::dnn::WGPUBackendWrapper::copyToHost() @ 03277a66:0x5d68fd
> cv::dnn::dnn4_v20200310::Net::Impl::getBlob(cv::dnn::dnn4_v20200310::LayerPin const&) @ 03277a66:0x4a6759
> cv::dnn::dnn4_v20200310::Net::forward(std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&) @ 03277a66:0x4a5134
> Wrappers::dnn_Net_forward_wrapper_1(cv::dnn::dnn4_v20200310::Net&) @ 03277a66:0x471ad
> emscripten::internal::FunctionInvoker<cv::Mat (*)(cv::dnn::dnn4_v20200310::Net&), cv::Mat, cv::dnn::dnn4_v20200310::Net&>::invoke(cv::Mat (**)(cv::dnn::dnn4_v20200310::Net&), cv::dnn::dnn4_v20200310::Net*) @ 03277a66:0x7702b
> dynCall_iii @ 03277a66:0x9ae3b7
> ret.<computed> @ opencv.js:8649
> Module.dynCall_iii @ opencv.js:9033
> dynCall_iii_1083 @ VM9080:4
> dnn_Net$forward @ VM9082:8
> proto.<computed> @ opencv.js:5754
> testCaffeLayer @ perf_conv2d.js:83
> async function (async)
> testCaffeLayer @ perf_conv2d.js:75
> runButton.onclick @ perf_conv2d.js:122
> ```

I test with _malloc and console the mappedData in JS for validation: [buffer.cpp](https://github.com/opencv/opencv/pull/18066/files#diff-6cbe7015aa6e3134d92db0b53f8c8e5fR11-R27)
The console log is 
```
Time cost(ms) : 8.54000000981614
perf_conv2d.js:86 null
perf_conv2d.js:98 Time1 cost(ms) : 2.490495049191805
perf_conv2d.js:99 Uint8Array(406800) [171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, …]
perf_conv2d.js:100 Net compute succeed
opencv.js:1808 sizeof arrayBuffer 406800
opencv.js:1808 Array buffer in JS Uint8Array(406800) [171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, 171, 170, 42, 62, …]
```
The buffer should be read correctly in JS, but it seems that the data is still not returned to the C code.
In addition, the order of the console does not seem to be as expected. It is expected that the buffer in JS should be consoled first. And how to free the memory allocated by _malloc in this part of JS ?
Is it the problem of my EM_JS code? Can you help me with suggestions? @huningxin

huningxin commented:
> In addition, the order of the console does not seem to be as expected. It is expected that the buffer in JS should be consoled first.

We may hit an [issue](https://github.com/emscripten-core/emscripten/issues/11890) of Asyncify and embind. I comment our findings there.

> And how to free the memory allocated by _malloc in this part of JS ?

You can free that memory in your C++ code.

huningxin commented:
After discuss with @NALLEIN and @wzw-intel , we agree to complete following tasks before the GSoC deadline:

For path native path (dawn):
- [ ] add a doc about how to build against dawn
- [ ] remove the dawn headers and binaries
- [ ] make sure the native path is buildable and testable

For js path:
- [ ] clean the js build script to make the js pass
- [ ] add a js test to prove the buffer content is readable but synchronization doesn't work due to an emscripten [issue](https://github.com/emscripten-core/emscripten/issues/11890)

For build bot:
- [ ] resolve the conflicts
- [ ] fix the bots failures

huningxin commented:
> > In addition, the order of the console does not seem to be as expected. It is expected that the buffer in JS should be consoled first.
> 
> We may hit an [issue](https://github.com/emscripten-core/emscripten/issues/11890) of Asyncify and embind. I comment our findings there.

@NALLEIN , I have a workaround for that issue. Please check it out and see whether you can apply that.

huningxin commented:
@NALLEIN , I put together the workaround in https://github.com/NALLEIN/opencv/pull/1. Please take a look and try on your side.

NALLEIN commented:
[GSoC report](https://gist.github.com/NALLEIN/d7e0357574a958b7a8edb442e3488271)

NALLEIN commented:
@alalek @huningxin It seems that the bot use the Emscripten version 1.38.21 but version  1.40.1 is required to build WebGPU. The version used by the CI bot is too low, causing an error when automatically build OpenCV.js and the check failed.
```
In file included from /build/precommit_docs/opencv/modules/dnn/src/webgpu/src/op_softmax.cpp:2:
/build/precommit_docs/opencv/modules/dnn/src/webgpu/src/common.hpp:15:10: fatal error: 'webgpu/webgpu_cpp.h' file not found
#include <webgpu/webgpu_cpp.h>
         ^~~~~~~~~~~~~~~~~~~~~
1 error generated.
shared:ERROR: compiler frontend failed to generate LLVM bitcode, halting
modules/dnn/CMakeFiles/opencv_dnn.dir/build.make:1413: recipe for target 'modules/dnn/CMakeFiles/opencv_dnn.dir/src/webgpu/src/op_softmax.cpp.o' failed
make[3]: *** [modules/dnn/CMakeFiles/opencv_dnn.dir/src/webgpu/src/op_softmax.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
```
How to fix this issue?

huningxin commented:
> How to fix this issue?

It needs @alalek 's help to upgrade the compiler version. At meanwhile, @NALLEIN we should make sure the build and test passed on local setup.

huningxin commented on 3rdparty/include/webgpu/README.md:
```diff
@@ -0,0 +1,81 @@
+## Build dawn to enable native implementation for DNN_BACKEND_WGPU  in modules/dnn
+
+
+### Build dawn
```
Please document the commit of dawn that you tested.

huningxin commented on platforms/js/build_js.py:
```diff
@@ -161,6 +161,7 @@ def get_cmake_cmd(self):
             cmd.append("-DBUILD_WASM_INTRIN_TESTS=ON")
         else:
             cmd.append("-DBUILD_WASM_INTRIN_TESTS=OFF")
+        cmd.append("-DWITH_WEBGPU=ON")
```
Please add an option for webgpu build. The default build should be without webgpu support.

huningxin commented on modules/js/src/core_bindings.cpp:
```diff
@@ -443,9 +448,10 @@ EMSCRIPTEN_BINDINGS(binding_utils)
         .constructor<>()
         .constructor<const Mat&>()
-        .constructor<Size, int>()
+        // .constructor<Size, int>()
```
Please avoid removing this constructor.

huningxin commented on modules/js/src/core_bindings.cpp:
```diff
@@ -443,9 +448,10 @@ EMSCRIPTEN_BINDINGS(binding_utils)
+        // .constructor<Size, int>()
         .constructor<int, int, int>()
         .constructor<int, int, int, const Scalar&>()
+        .constructor(&binding_utils::createMatWithDims, allow_raw_pointers())
```
If there is a conflict with other constructors, please try class function instead.

huningxin commented on modules/js/CMakeLists.txt:
```diff
@@ -101,6 +101,7 @@ endif()
 if(COMPILE_FLAGS)
     set_target_properties(${the_module} PROPERTIES COMPILE_FLAGS ${COMPILE_FLAGS})
 endif()
+set(EMSCRIPTEN_LINK_FLAGS "${EMSCRIPTEN_LINK_FLAGS} -s USE_WEBGPU=1 -s ASYNCIFY -s EXTRA_EXPORTED_RUNTIME_METHODS=['Asyncify']")
```
Please webgpu to guard these compiler switches.

huningxin commented on modules/js/test/test_dnn.js:
```diff
@@ -0,0 +1,126 @@
+    net1 = cv.readNetFromCaffe(path.modelPath, '');
+    net1.setInput(input);
+    const start1 = performance.now();
+    const out1 = net1.forward();
```
Please test the CPU backend first and assert its result.

huningxin reviewed (commented):
Thanks @NALLEIN , please share the test results for
1. native with WebGPU
2. native without WebGPU
3. web with WebGPU
4. web without WebGPU

huningxin commented:
> It needs @alalek 's help to upgrade the compiler version. At meanwhile, @NALLEIN we should make sure the build and test passed on local setup.

And please make the WebGPU build as optional, it would not break the default build.

huningxin commented:
> GSoC report

@NALLEIN , some feedbacks:
1. please summarize the shader modifications
2. add a table of the results of native tests on macOS and Linux (with dawn version)
3. add the step of build js tests and potential webgpu build option
4. add the js test and performance results with Chrome version
5. reorder the future work

NALLEIN commented:
> Thanks @NALLEIN , please share the test results for
> 
> 1. native with WebGPU
macOS : 10.15
Dawn comment  : 0b89b27263b127a46371641b299fdf0f2cd7caff

<img width="1269" alt="Screen Shot 2020-09-01 at 11 16 46 AM" src="https://user-images.githubusercontent.com/31888869/91793329-769da780-ec4a-11ea-8d08-a3a5409a8d33.png">

> 2. native without WebGPU
macOS : 10.15
Dawn comment  : 0b89b27263b127a46371641b299fdf0f2cd7caff

<img width="1278" alt="Screen Shot 2020-09-01 at 12 00 14 PM" src="https://user-images.githubusercontent.com/31888869/91793337-7b625b80-ec4a-11ea-992f-d3e454af9646.png">

> 3. web with WebGPU
> 4. web without WebGPU

NALLEIN commented on cmake/OpenCVDetectWebGPU.cmake:
```diff
@@ -0,0 +1,31 @@
+  set(DAWN_METAL 1)
+endif()
+
+if(NOT EMSCRIPTEN AND HAVE_WEBGPU)
```
Yes, this is the corret flag according to https://github.com/kainino0x/webgpu-cross-platform-demo/blob/main/CMakeLists.txt#L7

NALLEIN commented on modules/dnn/include/opencv2/dnn/dnn.hpp:
```diff
@@ -74,6 +74,7 @@ CV__DNN_INLINE_NS_BEGIN
         DNN_BACKEND_OPENCV,
         DNN_BACKEND_VKCOM,
         DNN_BACKEND_CUDA,
+        DNN_BACKEND_WGPU,
```
WGPU is the same as the Dawn's namespace `wgpu::`.

huningxin replied:
WEBGPU is clearer for backend name. Please use that.

NALLEIN commented on modules/dnn/src/webgpu/shader/concat_spv.cpp:
```diff
@@ -0,0 +1,83 @@
+
+extern const unsigned int concat_spv[541] = {
+    0x07230203,0x00010000,0x00080009,0x0000004b,0x00000000,0x00020011,0x00000001,0x0006000b,
+    0x00000001,0x4c534c47,0x6474732e,0x3035342e,0x00000000,0x0003000e,0x00000000,0x00000001,
```
Yes, these are pre-compiled shaders in [SPIR-V format](https://www.khronos.org/spir/). We can generate the SPIR-V shaders through the `spirv_generator.py` in `modules/dnn/src/webgpu/shader`.

huningxin replied:
I believe vkcom backend also uses this way, does it @wzw-intel ?

wzw-intel replied:
Yes.

NALLEIN commented on 3rdparty/include/webgpu/README.md:
```diff
@@ -0,0 +1,81 @@
+
+### Build dawn
+
+ Refer to [Dawn's build instructions](https://dawn.googlesource.com/dawn/+/HEAD/docs/buiding.md)  to complete the compilation of Dawn, and the directory structure is retained in opencv/3rdparty/include/webgpu to facilitate subsequent copying of the required files.
```
Sorry， I did not express it clearly， the retained folder structure means s[tandalone directory for preparing 3rdparty dependency](https://github.com/opencv/opencv/pull/18066/files/dda139feb95749dc8e7e65b38580fd0a0408f2f2#diff-96845957b59dcb213b5de07e800dd2f6R11).

NALLEIN commented:
[GSoC report](https://gist.github.com/NALLEIN/d7e0357574a958b7a8edb442e3488271)
[Demo](https://nallein.github.io/opencvjs/index.html) (layer test, requires the latest chrome dev/canary on macOS/windows and enable chrome://flags Unsafe WebGPU)
[Video](https://youtu.be/ZjOEMzp-0ys)

huningxin commented on 3rdparty/include/webgpu/README.md:
```diff
@@ -0,0 +1,30 @@
+
+### Update Dawn's API
+
+Dawn is still under development: [implementation status](https://github.com/gpuweb/gpuweb/wiki/Implementation-Status). If Dawn’s API changes, we have to re-build Dawn and copy the `webgpu_cpp.cpp` file under Dawn’s **out/Release/gen/src/dawn** folder to the `webgpu_cpp.cpp` file in OpenCV’s **modules/dnn/src/webgpu/src** folder. Only copy the content in `namespce wgpu {}`, no need to change the ` ...
```
Could you please avoid copying `webgpu_cpp.cpp` into `modules/dnn/src/webgpu/src`? Instead you can add this file into `SOURCE` of `dnn` module when `WITH_WEBGPU` is set. Please look into `modules/dnn/CMakeLists.txt` and `ocv_glob_module_sources` for details. 

@alalek , any comments here?

NALLEIN replied:
Done, I've moved this into `modules/dnn/CMakeLists.txt` and updated READEME.md. PTAL @huningxin @alalek !

huningxin replied:
Looks good, thanks!

NALLEIN commented:
@alalek  I've modified the cmake files and added OpenCV short license header into all the .h and .cpp files . 
The BuildBot has some warnings because the shader is compiled into the SPIR-V format. (Vulkan Backend uses the same method. The SPIR-V shader can be generated by the spirv_generator in the `modules/dd/src/webgpu/shader` directory. py to generate).
In addition, there is a failure because `FATAL: Build image for'javascript-webgpu--1.40.1' is missing on the current build worker` can you help to solve it ?

alalek commented:
> Build image for 'javascript-webgpu--1.40.1' is missing

There are some issues with preparing build image. I will return to this task on next week.

> BuildBot has some warnings

Warning is about the patch size - it is huge (about 50Mb). Please squash commits into one to remove unnecessary changes from git history (like dropped .so files).

NALLEIN commented:
> Warning is about the patch size - it is huge (about 50Mb). Please squash commits into one to remove unnecessary changes from git history (like dropped .so files).

Sorry, but I merged the upstream/master branch during development, which caused my commit to be mixed with others. I tried to rebase to squash commit but failed. Any  comments @huningxin @wzw-intel  ？

alalek commented:
> I tried to rebase to squash commit but failed.

It is due to "merges" in history (general suggestion is to avoid them, because they make git history unclear and break rebases). Prefer to use "git rebase" instead of "git merge"

I prepared rebased branch here (w/o git commit messages): https://github.com/alalek/opencv/commits/pr18066_r

NALLEIN commented:
> I prepared rebased branch here (w/o git commit messages): https://github.com/alalek/opencv/commits/pr18066_r

Thanks, I've squash the commits, but the BuildBot still has warnings :
 ```
emcc: warning: cannot represent a NaN literal '0x835e780' with custom bit pattern in NaN-canonicalizing JS engines (e.g. Firefox and Safari) without erasing bits!
emcc: warning: cannot represent a NaN literal '0x835e780' with custom bit pattern in NaN-canonicalizing JS engines (e.g. Firefox and Safari) without erasing bits!
```
![warning](https://user-images.githubusercontent.com/31888869/93765141-f8c82d00-fc46-11ea-912a-0c807f575c91.png)
And the could please help to resolve the custom build failure caused by `FATAL: Build image for 'javascript-webgpu--1.40.1' is missing on the current build worker` ?
![status](https://user-images.githubusercontent.com/31888869/93765154-fcf44a80-fc46-11ea-9f75-80df86cf6ed9.png)
@alalek .

NALLEIN commented:
@alalek PTAL the emcc warning has beed fixed, any more comments for this PR?

alalek reviewed (commented):
> --build_webgpu

This doesn't look to work properly.

CMake emits warning:

```
CMake Warning at cmake/OpenCVDetectWebGPU.cmake:19 (message):
  Can't use WebGPU-Dawn
```

Binary files are empty:

```
$ ll modules/dnn/CMakeFiles/opencv_dnn.dir/src/webgpu/src
total 60
drwxr-xr-x 2 build build 4096 Nov  5 21:02 ./
drwxr-xr-x 5 build build 4096 Nov  5 21:00 ../
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_buffer.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_concat.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_context.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_conv.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_internal.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_lrn.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_op_base.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_permute.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_pool.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_prior_box.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_relu.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_softmax.cpp.o
-rw-r--r-- 1 build build  260 Nov  5 21:02 webgpu_tensor.cpp.o
```

NALLEIN commented on modules/dnn/src/webgpu/README.md:
```diff
@@ -0,0 +1,26 @@
+export OPENCV_TEST_DATA_PATH=${PATH_TO}/opencv_extra/testdata
+
+sudo apt-get install libvulkan1 mesa-vulkan-drivers vulkan-utils
+git clone https://github.com/NALLEIN/opencv.git
+cd opencv/
+git checkout -b DawnTest origin/gsoc_2020_webgpu
```
> > --build_webgpu
> 
> This doesn't look to work properly.
> 
> CMake emits warning:
> 
> ```
> CMake Warning at cmake/OpenCVDetectWebGPU.cmake:19 (message):
>   Can't use WebGPU-Dawn
> ```
> 

Did you [pre-build the Dawn project](https://dawn.googlesource.com/dawn/+/HEAD/docs/buiding.md)? This PR has third-party dependencies and temporarily work with Dawn comment : `0b89b27263b127a46371641b299fdf0f2cd7caff`.

alalek replied:
Actually, I skipped the first section of PR's description with native code and `dawn` project.
These steps doesn't use Emscripten and doesn't produce opencv.js with WebGPU support:
```
cmake -D CMAKE_BUILD_TYPE=Release -DWITH_WEBGPU=ON -D CMAKE_INSTALL_PREFIX=/usr/local ..
make -j8
```

This comment is about the second section with `build_js.py --build_webgpu`.
Expected utilization of this: https://github.com/emscripten-core/emscripten/blob/master/system/include/webgpu/webgpu.h

/cc @huningxin What is expected to work in this patch?

alalek replied:
I checked `cvconfig.h` content from `build_js.py --build_webgpu` flow:
```
/* WebGPU-Dawn support */
/* #undef HAVE_WEBGPU */
#define DAWN_EMSDK
/* #undef DAWN_METAL */
```

Probably there is something wrong in CMake scripts.

NALLEIN replied:
> Actually, I skipped the first section of PR's description with native code and `dawn` project.
> These steps doesn't use Emscripten and doesn't produce opencv.js with WebGPU support:
> 
> ```
> cmake -D CMAKE_BUILD_TYPE=Release -DWITH_WEBGPU=ON -D CMAKE_INSTALL_PREFIX=/usr/local ..
> make -j8
> ```
> 
> This comment is about the second section with `build_js.py --build_webgpu`.
> Expected utilization of this: https://github.com/emscripten-core/emscripten/blob/master/system/include/webgpu/webgpu.h
> 
> /cc @huningxin What is expected to work in this patch?

To generate the corresponding binary file requires the first step, build native dawn. A third-party Dawn dependency is needed here, because WebGPU is implemented in native based on Dawn. 
Besides， I’m sorry I didn’t reply in time because of the school’s midterm exam.
@alalek

alalek replied:
According to [WebGPU demo](https://github.com/kainino0x/webgpu-cross-platform-demo) there are 2 cases:
- Native build (using dawn)
- Web build (to run in browser)

They are **independent**.

This thread is about "Web build" flavor.
Does it expected to work here?

NALLEIN replied:
> According to [WebGPU demo](https://github.com/kainino0x/webgpu-cross-platform-demo) there are 2 cases:
> 
> * Native build (using dawn)
> * Web build (to run in browser)
> 
> They are **independent**.
> 
> This thread is about "Web build" flavor.
> Does it expected to work here?

Yes, they are independent,  emscripten can map Dawn's C++ API to WebGPU's API. But build opencv.js will not generate binary files corresponding to cpp files.  The finally generated opencv.js file is in the opencv/build_js/bin directory.

stereomatchingkiss replied:
Do this mean it wouldn't generate library file could link by c++ program(ex : libopencv_dnn.a)?Would like to port the native app to browser in order to perform object detection/fall down classification.

asenyaev commented:
jenkins cn please retry a build

dmatveev commented:
This is actually quite cool.. Sad it is not merged & got outdated with all the DNN changes since then.
